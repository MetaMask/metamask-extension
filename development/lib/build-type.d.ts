import { Infer, Struct } from 'superstruct';

// TODO: Replace `any` with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Unique<Element extends Struct<any>> = (
  struct: Struct<Infer<Element>[], Infer<Element>>,
  eq?: (a: Infer<Element>, b: Infer<Element>) => boolean,
) => Struct<Infer<Element>[], Infer<Element>>;
