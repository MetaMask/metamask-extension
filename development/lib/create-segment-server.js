const http = require('http');

/**
 * This is the default error handler to be used by this mock segment server.
 * It will print the error to the console and exit the process.
 *
 * @param {Error} error - The server error
 */
function defaultOnError(error) {
  console.log(error);
  process.exit(1);
}

/**
 * @typedef {import('http').IncomingMessage} IncomingMessage
 * @typedef {import('http').ServerResponse} ServerResponse
 */

/**
 * This function handles requests for the mock Segment server
 *
 * @typedef {(request: IncomingMessage, response: ServerResponse, metricEvents: Array<object>) => void} MockSegmentRequestHandler
 */

/**
 * Creates a HTTP server that acts as a fake version of the Segment API.
 * The bevahiour is rudimentary at the moment - it returns HTTP 200 in response
 * to every request. The only function this serves is to spy on requests sent to
 * this server, and to parse the request payloads as Segment batch events.
 *
 * @param {MockSegmentRequestHandler} onRequest - A callback for each request the server receives.
 * @param {(error: Error) => void} [onError] - A callback for server error events
 */
function createSegmentServer(onRequest, onError = defaultOnError) {
  const server = http.createServer(async (request, response) => {
    const chunks = [];

    request.on('data', (chunk) => {
      chunks.push(chunk);
    });

    await new Promise((resolve) => {
      request.on('end', () => {
        resolve();
      });
    });

    // respond to preflight request
    if (request.method === 'OPTIONS') {
      response.setHeader('Access-Control-Allow-Origin', '*');
      response.setHeader('Access-Control-Allow-Methods', '*');
      response.setHeader('Access-Control-Allow-Headers', '*');
      response.statusCode = 200;
      response.end();
      return;
    }

    let metricEvents = [];
    if (chunks.length) {
      const body = Buffer.concat(chunks).toString();
      const segmentPayload = JSON.parse(body);
      metricEvents = segmentPayload.batch;
    }

    onRequest(request, response, metricEvents);
  });

  server.on('error', onError);

  return {
    start: async (port) => {
      await new Promise((resolve, reject) => {
        server.listen(port, (error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
      });
    },
    stop: async () => {
      await new Promise((resolve, reject) => {
        server.close((error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
        // We need to close all connections to stop the server quickly
        // Otherwise it takes a few seconds for it to close
        server.closeAllConnections();
      });
    },
  };
}

module.exports = { createSegmentServer };
