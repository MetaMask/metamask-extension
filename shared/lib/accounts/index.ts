export * from './accounts';
export * from './bitcoin-wallet-snap';
export * from './snaps';
export * from './solana-wallet-snap';
