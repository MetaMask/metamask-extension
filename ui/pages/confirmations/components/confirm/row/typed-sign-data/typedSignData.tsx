import React from 'react';

import { useI18nContext } from '../../../../../../hooks/useI18nContext';
import { Box } from '../../../../../../components/component-library';
import { BlockSize } from '../../../../../../helpers/constants/design-system';
import {
  ConfirmInfoRow,
  ConfirmInfoRowText,
} from '../../../../../../components/app/confirm/info/row';
import { parseSanitizeTypedDataMessage } from '../../../../utils';
import { DataTree } from '../dataTree';

export const ConfirmInfoRowTypedSignData = ({
  data,
  tokenDecimals,
  chainId,
}: {
  data: string;
  isPermit?: boolean;
  tokenDecimals?: number;
  chainId: string;
}) => {
  const t = useI18nContext();

  if (!data) {
    return null;
  }

  const { sanitizedMessage, primaryType } = parseSanitizeTypedDataMessage(data);

  return (
    <Box width={BlockSize.Full}>
      <ConfirmInfoRow
        label={`${t('primaryType')}:`}
        style={{ paddingLeft: 0, paddingRight: 0 }}
      >
        <ConfirmInfoRowText text={primaryType} />
      </ConfirmInfoRow>
      <Box style={{ marginLeft: -8 }}>
        <DataTree
          data={sanitizedMessage.value}
          primaryType={primaryType}
          tokenDecimals={tokenDecimals}
          chainId={chainId}
        />
      </Box>
    </Box>
  );
};
