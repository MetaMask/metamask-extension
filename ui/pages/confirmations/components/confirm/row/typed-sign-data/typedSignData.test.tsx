import React from 'react';
import {
  rawMessageV4,
  unapprovedTypedSignMsgV4,
} from '../../../../../../../test/data/confirmations/typed_sign';
import mockState from '../../../../../../../test/data/mock-state.json';
import { renderWithProvider } from '../../../../../../../test/lib/render-helpers';
import configureStore from '../../../../../../store/store';
import { ConfirmInfoRowTypedSignData } from './typedSignData';

const CHAIN_ID_MOCK = '0x123';

describe('ConfirmInfoRowTypedSignData', () => {
  const renderWithComponentData = (
    data = unapprovedTypedSignMsgV4.msgParams?.data as string,
  ) => {
    const store = configureStore(mockState);

    return renderWithProvider(
      <ConfirmInfoRowTypedSignData data={data} chainId={CHAIN_ID_MOCK} />,
      store,
    );
  };

  it('should match snapshot', () => {
    const { container } = renderWithComponentData(
      unapprovedTypedSignMsgV4.msgParams?.data as string,
    );
    expect(container).toMatchSnapshot();
  });

  it('should return null if data is not defined', () => {
    const { container } = renderWithComponentData('');
    expect(container).toBeEmptyDOMElement();
  });

  it('should not render data whose type is not defined', () => {
    // TODO: Replace `any` with type
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const mockRawMessageV4 = { ...rawMessageV4 } as any;

    mockRawMessageV4.message.do_not_display = 'one';
    mockRawMessageV4.message.do_not_display_2 = {
      do_not_display: 'two',
    };

    const mockV4MsgParamsData = JSON.stringify(mockRawMessageV4);
    const { queryByText } = renderWithComponentData(mockV4MsgParamsData);

    expect(queryByText('do_not_display')).not.toBeInTheDocument();
    expect(queryByText('one')).not.toBeInTheDocument();
    expect(queryByText('do_not_display_2')).not.toBeInTheDocument();
    expect(queryByText('two')).not.toBeInTheDocument();
  });
});
