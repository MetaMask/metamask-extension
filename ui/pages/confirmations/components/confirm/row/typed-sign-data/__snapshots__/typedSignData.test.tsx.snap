// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmInfoRowTypedSignData should match snapshot 1`] = `
<div>
  <div
    class="mm-box mm-box--width-full"
  >
    <div
      class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
      style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-left: 0px; padding-right: 0px;"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--align-items-center"
        >
          <p
            class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
          >
            Primary type:
          </p>
        </div>
      </div>
      <div
        class="mm-box mm-box--display-flex mm-box--gap-2 mm-box--flex-wrap-wrap mm-box--align-items-center mm-box--min-width-0"
      >
        <p
          class="mm-box mm-text mm-text--body-md mm-box--color-inherit"
          style="white-space: pre-wrap;"
        >
          Mail
        </p>
      </div>
    </div>
    <div
      class="mm-box"
      style="margin-left: -8px;"
    >
      <div
        class="mm-box mm-box--width-full"
      >
        <div
          class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
          style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
        >
          <div
            class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
          >
            <div
              class="mm-box mm-box--display-flex mm-box--align-items-center"
            >
              <p
                class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
              >
                Contents:
              </p>
            </div>
          </div>
          <div
            class="mm-box mm-box--display-flex mm-box--gap-2 mm-box--flex-wrap-wrap mm-box--align-items-center mm-box--min-width-0"
          >
            <p
              class="mm-box mm-text mm-text--body-md mm-box--color-inherit"
              style="white-space: pre-wrap;"
            >
              Hello, Bob!
            </p>
          </div>
        </div>
        <div
          class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
          style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
        >
          <div
            class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
          >
            <div
              class="mm-box mm-box--display-flex mm-box--align-items-center"
            >
              <p
                class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
              >
                From:
              </p>
            </div>
          </div>
          <div
            class="mm-box mm-box--width-full"
          >
            <div
              class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
              style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
            >
              <div
                class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
              >
                <div
                  class="mm-box mm-box--display-flex mm-box--align-items-center"
                >
                  <p
                    class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                  >
                    Name:
                  </p>
                </div>
              </div>
              <div
                class="mm-box mm-box--display-flex mm-box--gap-2 mm-box--flex-wrap-wrap mm-box--align-items-center mm-box--min-width-0"
              >
                <p
                  class="mm-box mm-text mm-text--body-md mm-box--color-inherit"
                  style="white-space: pre-wrap;"
                >
                  Cow
                </p>
              </div>
            </div>
            <div
              class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
              style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
            >
              <div
                class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
              >
                <div
                  class="mm-box mm-box--display-flex mm-box--align-items-center"
                >
                  <p
                    class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                  >
                    Wallets:
                  </p>
                </div>
              </div>
              <div
                class="mm-box mm-box--width-full"
              >
                <div
                  class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                  style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--align-items-center"
                    >
                      <p
                        class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                      >
                        0:
                      </p>
                    </div>
                  </div>
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                    >
                      <div
                        class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-account mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
                      >
                        <div
                          class="mm-avatar-account__jazzicon"
                        >
                          <div
                            style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 16px; height: 16px; display: inline-block; background: rgb(35, 98, 225);"
                          >
                            <svg
                              height="16"
                              width="16"
                              x="0"
                              y="0"
                            >
                              <rect
                                fill="#FA7900"
                                height="16"
                                transform="translate(-0.1251869550000172 -0.18581992328639055) rotate(237.8 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                              <rect
                                fill="#034F5E"
                                height="16"
                                transform="translate(5.484089776243406 -5.806818790918918) rotate(321.2 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                              <rect
                                fill="#FB185C"
                                height="16"
                                transform="translate(4.337646399293827 10.83159092527544) rotate(190.8 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                      <p
                        class="mm-box mm-text mm-text--body-md mm-box--margin-left-2 mm-box--color-inherit"
                        data-testid="confirm-info-row-display-name"
                      >
                        0xCD2a3...DD826
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                  style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--align-items-center"
                    >
                      <p
                        class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                      >
                        1:
                      </p>
                    </div>
                  </div>
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                    >
                      <div
                        class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-account mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
                      >
                        <div
                          class="mm-avatar-account__jazzicon"
                        >
                          <div
                            style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 16px; height: 16px; display: inline-block; background: rgb(24, 183, 242);"
                          >
                            <svg
                              height="16"
                              width="16"
                              x="0"
                              y="0"
                            >
                              <rect
                                fill="#F91E01"
                                height="16"
                                transform="translate(-3.031601448459508 1.44638157917772) rotate(179.5 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                              <rect
                                fill="#FA5300"
                                height="16"
                                transform="translate(0.001001577275992914 -7.592133809870717) rotate(370.7 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                              <rect
                                fill="#F5A300"
                                height="16"
                                transform="translate(-9.667744843805323 10.267175012988094) rotate(298.9 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                      <p
                        class="mm-box mm-text mm-text--body-md mm-box--margin-left-2 mm-box--color-inherit"
                        data-testid="confirm-info-row-display-name"
                      >
                        0xDeaDb...DbeeF
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                  style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--align-items-center"
                    >
                      <p
                        class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                      >
                        2:
                      </p>
                    </div>
                  </div>
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                    >
                      <div
                        class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-account mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
                      >
                        <div
                          class="mm-avatar-account__jazzicon"
                        >
                          <div
                            style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 16px; height: 16px; display: inline-block; background: rgb(245, 143, 0);"
                          >
                            <svg
                              height="16"
                              width="16"
                              x="0"
                              y="0"
                            >
                              <rect
                                fill="#FA3E00"
                                height="16"
                                transform="translate(2.1189727229825075 3.7201589705667972) rotate(163.6 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                              <rect
                                fill="#018E74"
                                height="16"
                                transform="translate(-3.5064720430845115 9.598091001429925) rotate(205.4 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                              <rect
                                fill="#238FE1"
                                height="16"
                                transform="translate(-5.001213401730692 12.04822844354251) rotate(286.5 8 8)"
                                width="16"
                                x="0"
                                y="0"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                      <p
                        class="mm-box mm-text mm-text--body-md mm-box--margin-left-2 mm-box--color-inherit"
                        data-testid="confirm-info-row-display-name"
                      >
                        0x06195...43896
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
          style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
        >
          <div
            class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
          >
            <div
              class="mm-box mm-box--display-flex mm-box--align-items-center"
            >
              <p
                class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
              >
                To:
              </p>
            </div>
          </div>
          <div
            class="mm-box mm-box--width-full"
          >
            <div
              class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
              style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
            >
              <div
                class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
              >
                <div
                  class="mm-box mm-box--display-flex mm-box--align-items-center"
                >
                  <p
                    class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                  >
                    0:
                  </p>
                </div>
              </div>
              <div
                class="mm-box mm-box--width-full"
              >
                <div
                  class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                  style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--align-items-center"
                    >
                      <p
                        class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                      >
                        Name:
                      </p>
                    </div>
                  </div>
                  <div
                    class="mm-box mm-box--display-flex mm-box--gap-2 mm-box--flex-wrap-wrap mm-box--align-items-center mm-box--min-width-0"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-md mm-box--color-inherit"
                      style="white-space: pre-wrap;"
                    >
                      Bob
                    </p>
                  </div>
                </div>
                <div
                  class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                  style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                  >
                    <div
                      class="mm-box mm-box--display-flex mm-box--align-items-center"
                    >
                      <p
                        class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                      >
                        Wallets:
                      </p>
                    </div>
                  </div>
                  <div
                    class="mm-box mm-box--width-full"
                  >
                    <div
                      class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                      style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                    >
                      <div
                        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                      >
                        <div
                          class="mm-box mm-box--display-flex mm-box--align-items-center"
                        >
                          <p
                            class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                          >
                            0:
                          </p>
                        </div>
                      </div>
                      <div
                        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                      >
                        <div
                          class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                        >
                          <div
                            class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-account mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
                          >
                            <div
                              class="mm-avatar-account__jazzicon"
                            >
                              <div
                                style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 16px; height: 16px; display: inline-block; background: rgb(242, 162, 2);"
                              >
                                <svg
                                  height="16"
                                  width="16"
                                  x="0"
                                  y="0"
                                >
                                  <rect
                                    fill="#C8144A"
                                    height="16"
                                    transform="translate(3.3716958087807476 -1.5355449411453455) rotate(359.3 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                  <rect
                                    fill="#01878E"
                                    height="16"
                                    transform="translate(-4.56026532072257 -4.118767169186573) rotate(335.8 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                  <rect
                                    fill="#1897F2"
                                    height="16"
                                    transform="translate(-12.66850392914138 2.191919563316492) rotate(275.0 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                </svg>
                              </div>
                            </div>
                          </div>
                          <p
                            class="mm-box mm-text mm-text--body-md mm-box--margin-left-2 mm-box--color-inherit"
                            data-testid="confirm-info-row-display-name"
                          >
                            0xbBbBB...bBBbB
                          </p>
                        </div>
                      </div>
                    </div>
                    <div
                      class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                      style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                    >
                      <div
                        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                      >
                        <div
                          class="mm-box mm-box--display-flex mm-box--align-items-center"
                        >
                          <p
                            class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                          >
                            1:
                          </p>
                        </div>
                      </div>
                      <div
                        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                      >
                        <div
                          class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                        >
                          <div
                            class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-account mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
                          >
                            <div
                              class="mm-avatar-account__jazzicon"
                            >
                              <div
                                style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 16px; height: 16px; display: inline-block; background: rgb(250, 54, 0);"
                              >
                                <svg
                                  height="16"
                                  width="16"
                                  x="0"
                                  y="0"
                                >
                                  <rect
                                    fill="#2395E1"
                                    height="16"
                                    transform="translate(-1.7353143855136905 0.5115867848025443) rotate(191.9 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                  <rect
                                    fill="#F90101"
                                    height="16"
                                    transform="translate(1.7169232764557802 -9.276715881533601) rotate(319.8 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                  <rect
                                    fill="#F58700"
                                    height="16"
                                    transform="translate(9.91414735179574 5.139720751783702) rotate(180.7 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                </svg>
                              </div>
                            </div>
                          </div>
                          <p
                            class="mm-box mm-text mm-text--body-md mm-box--margin-left-2 mm-box--color-inherit"
                            data-testid="confirm-info-row-display-name"
                          >
                            0xB0Bda...bEa57
                          </p>
                        </div>
                      </div>
                    </div>
                    <div
                      class="mm-box confirm-info-row mm-box--margin-top-2 mm-box--margin-bottom-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--flex-direction-row mm-box--flex-wrap-wrap mm-box--justify-content-space-between mm-box--align-items-center mm-box--color-text-default mm-box--rounded-lg"
                      style="overflow-wrap: anywhere; min-height: 24px; position: relative; padding-right: 0px;"
                    >
                      <div
                        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--align-items-flex-start"
                      >
                        <div
                          class="mm-box mm-box--display-flex mm-box--align-items-center"
                        >
                          <p
                            class="mm-box mm-text mm-text--body-md-medium mm-box--color-inherit"
                          >
                            2:
                          </p>
                        </div>
                      </div>
                      <div
                        class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                      >
                        <div
                          class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                        >
                          <div
                            class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-account mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
                          >
                            <div
                              class="mm-avatar-account__jazzicon"
                            >
                              <div
                                style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 16px; height: 16px; display: inline-block; background: rgb(200, 20, 86);"
                              >
                                <svg
                                  height="16"
                                  width="16"
                                  x="0"
                                  y="0"
                                >
                                  <rect
                                    fill="#018E8C"
                                    height="16"
                                    transform="translate(3.988074950687866 0.3554940504644344) rotate(177.8 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                  <rect
                                    fill="#F29202"
                                    height="16"
                                    transform="translate(-7.220928666254465 -2.4813555582770794) rotate(334.6 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                  <rect
                                    fill="#236FE1"
                                    height="16"
                                    transform="translate(14.621768557060026 -6.282465882779628) rotate(456.1 8 8)"
                                    width="16"
                                    x="0"
                                    y="0"
                                  />
                                </svg>
                              </div>
                            </div>
                          </div>
                          <p
                            class="mm-box mm-text mm-text--body-md mm-box--margin-left-2 mm-box--color-inherit"
                            data-testid="confirm-info-row-display-name"
                          >
                            0xB0B0b...00000
                          </p>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
